File: android/app/src/main/kotlin/dev/agentdeck/terrarium/CreatureLayout.kt
package dev.agentdeck.terrarium

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.sin

/**
 * Layout slot for positioning a creature in the terrarium.
 */
data class CreatureSlot(
    val centerXFraction: Float,
    val centerYFraction: Float,
    val scaleFactor: Float,
)

/**
 * Compute layout positions for multiple octopuses (coding agents).
 * Distributes them across the left-center area of the terrarium.
 */
fun layoutOctopuses(count: Int): List<CreatureSlot> {
    return when (count) {
        0 -> emptyList()
        1 -> listOf(
            CreatureSlot(
                TerrariumLayout.OCTOPUS_CENTER_X_FRACTION,
                TerrariumLayout.OCTOPUS_CENTER_Y_FRACTION,
                1.0f,
            )
        )
        2 -> listOf(
            CreatureSlot(0.30f, 0.42f, 0.85f),
            CreatureSlot(0.55f, 0.48f, 0.85f),
        )
        3 -> listOf(
            CreatureSlot(0.30f, 0.38f, 0.75f),
            CreatureSlot(0.52f, 0.38f, 0.75f),
            CreatureSlot(0.40f, 0.52f, 0.75f),
        )
        else -> {
            // Grid layout for 4+, shrinking as needed
            val scale = max(0.5f, 0.75f - (count - 3) * 0.05f)
            val cols = if (count <= 4) 2 else 3
            val rows = (count + cols - 1) / cols
            val startX = 0.20f
            val endX = 0.60f
            val startY = 0.32f
            val endY = 0.55f
            val dx = if (cols > 1) (endX - startX) / (cols - 1) else 0f
            val dy = if (rows > 1) (endY - startY) / (rows - 1) else 0f

            (0 until count).map { i ->
                val col = i % cols
                val row = i / cols
                CreatureSlot(
                    startX + col * dx,
                    startY + row * dy,
                    scale,
                )
            }
        }
    }
}

/**
 * Compute layout positions for OpenClaw worker crayfish.
 * Workers are smaller and arranged in an arc around the main crayfish position.
 */
fun layoutWorkerCrayfish(count: Int): List<CreatureSlot> {
    if (count == 0) return emptyList()

    val mainX = TerrariumLayout.CRAYFISH_CENTER_X_FRACTION
    val mainY = TerrariumLayout.CRAYFISH_CENTER_Y_FRACTION
    val arcRadius = 0.08f

    return (0 until count).map { i ->
        val angle = PI.toFloat() * 0.8f + (i.toFloat() / max(1, count - 1).toFloat()) * PI.toFloat() * 0.4f
        CreatureSlot(
            mainX + cos(angle) * arcRadius,
            mainY + sin(angle) * arcRadius,
            0.5f,
        )
    }
}

File: apple/AgentDeck/Terrarium/CreatureLayout.swift
// CreatureLayout.swift — Multi-session creature positioning
// Ported from android CreatureLayout.kt

import Foundation

struct CreatureSlot {
    let x: Float
    let y: Float
    let scale: Float
}

enum CreatureLayout {
    /// Layout octopus positions for N agents
    static func layoutOctopuses(count: Int) -> [CreatureSlot] {
        switch count {
        case 0:
            return []
        case 1:
            return [CreatureSlot(x: 0.4, y: 0.45, scale: 1.0)]
        case 2:
            return [
                CreatureSlot(x: 0.28, y: 0.38, scale: 0.85),
                CreatureSlot(x: 0.55, y: 0.50, scale: 0.85),
            ]
        case 3:
            return [
                CreatureSlot(x: 0.22, y: 0.35, scale: 0.75),
                CreatureSlot(x: 0.52, y: 0.35, scale: 0.75),
                CreatureSlot(x: 0.36, y: 0.52, scale: 0.75),
            ]
        default:
            return layoutGrid(count: count)
        }
    }

    static func layoutCloudCreatures(count: Int) -> [CreatureSlot] {
        switch count {
        case 0:
            return []
        case 1:
            return [CreatureSlot(x: 0.55, y: 0.20, scale: 1.0)]
        case 2:
            return [
                CreatureSlot(x: 0.40, y: 0.18, scale: 0.85),
                CreatureSlot(x: 0.62, y: 0.22, scale: 0.85),
            ]
        case 3:
            return [
                CreatureSlot(x: 0.35, y: 0.16, scale: 0.75),
                CreatureSlot(x: 0.55, y: 0.20, scale: 0.75),
                CreatureSlot(x: 0.45, y: 0.28, scale: 0.75),
            ]
        default:
            let cols = count <= 4 ? 2 : 3
            let rows = Int(ceil(Double(count) / Double(cols)))
            let scale = max(0.50, 0.75 - Float(count - 3) * 0.05)
            let startX: Float = 0.30
            let endX: Float = 0.65
            let startY: Float = 0.12
            let endY: Float = 0.30

            var slots: [CreatureSlot] = []
            for i in 0..<count {
                let col = i % cols
                let row = i / cols
                let x = startX + (cols > 1 ? Float(col) / Float(cols - 1) * (endX - startX) : (endX - startX) / 2)
                let y = startY + (rows > 1 ? Float(row) / Float(rows - 1) * (endY - startY) : (endY - startY) / 2)
                slots.append(CreatureSlot(x: x, y: y, scale: scale))
            }
            return slots
        }
    }

    static func layoutOpenCodeCreatures(count: Int) -> [CreatureSlot] {
        switch count {
        case 0:
            return []
        case 1:
            return [CreatureSlot(x: 0.48, y: 0.40, scale: 1.0)]
        case 2:
            return [
                CreatureSlot(x: 0.38, y: 0.38, scale: 0.85),
                CreatureSlot(x: 0.58, y: 0.42, scale: 0.85),
            ]
        case 3:
            return [
                CreatureSlot(x: 0.33, y: 0.36, scale: 0.75),
                CreatureSlot(x: 0.53, y: 0.36, scale: 0.75),
                CreatureSlot(x: 0.43, y: 0.48, scale: 0.75),
            ]
        default:
            let cols = count <= 4 ? 2 : 3
            let rows = Int(ceil(Double(count) / Double(cols)))
            let scale = max(0.50, 0.75 - Float(count - 3) * 0.05)
            let startX: Float = 0.28
            let endX: Float = 0.62
            let startY: Float = 0.32
            let endY: Float = 0.50

            var slots: [CreatureSlot] = []
            for i in 0..<count {
                let col = i % cols
                let row = i / cols
                let x = startX + (cols > 1 ? Float(col) / Float(cols - 1) * (endX - startX) : (endX - startX) / 2)
                let y = startY + (rows > 1 ? Float(row) / Float(rows - 1) * (endY - startY) : (endY - startY) / 2)
                slots.append(CreatureSlot(x: x, y: y, scale: scale))
            }
            return slots
        }
    }

    private static func layoutGrid(count: Int) -> [CreatureSlot] {
        let cols = count <= 4 ? 2 : 3
        let rows = Int(ceil(Double(count) / Double(cols)))
        let scale = max(0.45, 0.75 - Float(count - 3) * 0.05)

        // Respect swim boundaries: 0.20~0.62 X, 0.32~0.55 Y
        let startX: Float = 0.20
        let endX: Float = 0.62
        let startY: Float = 0.32
        let endY: Float = 0.55

        var slots: [CreatureSlot] = []
        for i in 0..<count {
            let col = i % cols
            let row = i / cols
            let x = startX + (cols > 1 ? Float(col) / Float(cols - 1) * (endX - startX) : (endX - startX) / 2)
            let y = startY + (rows > 1 ? Float(row) / Float(rows - 1) * (endY - startY) : (endY - startY) / 2)
            slots.append(CreatureSlot(x: x, y: y, scale: scale))
        }
        return slots
    }
}
